docs/develop/create-sso-office-add-ins-aspnet.md

```diff
@@ -31,536 +33,359 @@ Clone or download the repo at [Office Add-in ASPNET SSO](https://github.com/Offi
-
-    ```javascript
-    if ((exceptionMessage.indexOf("AADSTS500133") !== -1)
-        && (retryGetAccessToken <= 0)) {
+1. In the `handleServerSideErrors` function, replace `TODO 7` with the following code. About this code, note:
 
-        retryGetAccessToken++;
-        getGraphData();
-    }
-    ```
-
-1. Replace `TODO 8` with the following:
+    - In the rare case the original SSO token is expired, it will detect this error condition and call `getUserFilenames` again. This results in another call to `getAccessToken` which returns a refreshed access token. The `retryGetAccessToken` variable counts the retries and is currently configured to only retry once.
+    - Finally, if an error cannot be handled, the default is to display the error in the task pane.
 
     ```javascript
-    else {
-        dialogFallback();
+    // Results from other errors (other than AADSTS50076) will have an ExceptionMessage property.
+    const exceptionMessage = JSON.parse(errorResponse.responseText).ExceptionMessage;
+    if (exceptionMessage) {
+        // On rare occasions the access token is unexpired when Office validates it,
+        // but expires by the time it is sent to Microsoft identity in the OBO flow. Microsoft identity will respond
+        // with "The provided value for the 'assertion' is not valid. The assertion has expired."
+        // Retry the call of getAccessToken (no more than once). This time Office will return a 
+        // new unexpired access token.
+        if ((exceptionMessage.includes("AADSTS500133"))
+            && (retryGetAccessToken <= 0)) {
+            retryGetAccessToken++;
+            getUserFileNames();
+            return;
+        }
+        else {
+            showMessage("MSAL error from application server: " + JSON.stringify(exceptionMessage));
+            return;
+        }
     }
+    // Default error handling if previous checks didn't apply.
+    showMessage(errorResponse.responseJSON.value);
     ```
 
-1. Save the file.
-
 ## Code the server side
 
-### Configure the OWIN middleware
+The server-side code is an ASP.NET Core server that provides REST APIs for the client to call. For example, the REST API `/api/files` gets a list of filenames from the user's OneDrive folder. Each REST API call requires an access token by the client to ensure the correct client is accessing their data. The access token is exchanged for a Microsoft Graph token through the On-Behalf-Of flow (OBO). The new Microsoft Graph token is cached by the MSAL.NET library for subsequent API calls. It's never sent outside of the server-side code. Microsoft identity documentation refers to this server as the middle-tier server because it is in the middle of the flow from client-side code to Microsoft services. For more information, see [Middle-tier access token request](/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow#middle-tier-access-token-request)
```

Comment:
Thanks for adding this call out! Important security considerations for our customers! :)